flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java

```diff
@@ -0,0 +1,491 @@
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static org.apache.flink.core.testutils.FlinkMatchers.containsCause;
+import static org.apache.flink.runtime.scheduler.adaptive.ExecutingTest.createFailingStateTransition;
+import static org.apache.flink.runtime.scheduler.adaptive.WaitingForResourcesTest.assertNonNull;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/** Tests for the {@link StopWithSavepoint} state. */
+public class StopWithSavepointTest extends TestLogger {
+    private static final String SAVEPOINT_PATH = "test://savepoint/path";
+
+    @Test
+    public void testFinishedOnSuccessfulStopWithSavepoint() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StateTrackingMockExecutionGraph mockExecutionGraph =
+                    new StateTrackingMockExecutionGraph();
+            CompletableFuture<String> savepointFuture = new CompletableFuture<>();
+
+            StopWithSavepoint sws =
+                    createStopWithSavepoint(ctx, mockExecutionGraph, savepointFuture);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setExpectFinished(assertNonNull());
+
+            savepointFuture.complete(SAVEPOINT_PATH);
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FINISHED);
+            ctx.triggerScheduledExecutors();
+
+            assertThat(sws.getOperationFuture().get(), is(SAVEPOINT_PATH));
+        }
+    }
+
+    @Test
+    public void testJobFailed() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StateTrackingMockExecutionGraph mockExecutionGraph =
+                    new StateTrackingMockExecutionGraph();
+            StopWithSavepoint sws = createStopWithSavepoint(ctx, mockExecutionGraph);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+
+            ctx.setExpectFailing(
+                    failingArguments -> {
+                        assertThat(
+                                failingArguments.getExecutionGraph().getState(),
+                                is(JobStatus.FAILED));
+                        assertThat(
+                                failingArguments.getFailureCause(),
+                                containsCause(FlinkException.class));
+                    });
+
+            // fail job:
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FAILED);
+            ctx.triggerScheduledExecutors();
+
+            assertThat(sws.getOperationFuture().isCompletedExceptionally(), is(true));
+        }
```

Comment:
Does this test completes the savepoint operation? This would be good to know for the test whether the state waits for it or not.

Reply:
I'm not sure if I fully understood this comment. I've added `testJobFailedAndSavepointOperationFails`, hoping to address your concern.

Reply:
The question was whether the savepoint future is completed or not. Looking at this test, this is not clear to me. I guess this is somewhere hidden in the default values when creating `createStopWithSavepoint`.